s3-source-connector/src/main/java/io/aiven/kafka/connect/s3/source/input/ByteArrayTransformer.java

```diff
@@ -37,29 +39,32 @@ public void configureValueConverter(final Map<String, String> config, final S3So
         // For byte array transformations, ByteArrayConverter is the converter which is the default config.
     }
 
-    @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
     @Override
-    public List<Object> getRecords(final InputStream inputStream, final String topic, final int topicPartition,
+    public Stream<Object> getRecords(final InputStream inputStream, final String topic, final int topicPartition,
             final S3SourceConfig s3SourceConfig) {
-
         final int maxMessageBytesSize = s3SourceConfig.getInt(EXPECTED_MAX_MESSAGE_BYTES);
-        final byte[] buffer = new byte[maxMessageBytesSize];
-        int bytesRead;
 
-        final List<Object> chunks = new ArrayList<>();
-        try {
-            bytesRead = inputStream.read(buffer);
-            while (bytesRead != -1) {
-                final byte[] chunk = new byte[bytesRead];
-                System.arraycopy(buffer, 0, chunk, 0, bytesRead);
-                chunks.add(chunk);
-                bytesRead = inputStream.read(buffer);
-            }
-        } catch (IOException e) {
-            LOGGER.error("Error reading from input stream: {}", e.getMessage(), e);
-        }
+        // Create a Stream that processes each chunk lazily
+        return StreamSupport.stream(new Spliterators.AbstractSpliterator<>(Long.MAX_VALUE, Spliterator.ORDERED) {
+            final byte[] buffer = new byte[maxMessageBytesSize];
 
-        return chunks;
+            @Override
+            public boolean tryAdvance(final java.util.function.Consumer<? super Object> action) {
+                try {
+                    final int bytesRead = inputStream.read(buffer);
+                    if (bytesRead == -1) {
+                        return false;
+                    }
+                    final byte[] chunk = new byte[bytesRead];
+                    System.arraycopy(buffer, 0, chunk, 0, bytesRead);
+                    action.accept(chunk);
+                    return true;
+                } catch (IOException e) {
+                    LOGGER.error("Error reading from input stream: {}", e.getMessage(), e);
```

Comment:
Might be able to make this error message marginally more descriptive.

"Error trying to advance byte stream: {}" or something similar.